File: Python/Algorithm/Algorithm I/278. First Bad Version.py
class Solution:
    def firstBadVersion(self, n: int) -> int:
        """
        Find the first bad version in a series of versions using binary search.

        Args:
            n (int): The number of versions.

        Returns:
            int: The first bad version.

        Time Complexity: o(log(n)) where n is the number of versions.
        Space Complexity: o(1) since we are not using any additional space.
        """
        leftPointer = 1
        rightPointer = n

        while leftPointer <= rightPointer:
            middlePointer = leftPointer + (rightPointer - leftPointer) // 2

            if isBadVersion(middlePointer):
                rightPointer = middlePointer - 1
            else:
                leftPointer = middlePointer + 1

        return rightPointer + 1

File: C/Algorithm/Algorithm I/278. First Bad Version.c
int firstBadVersion(int n) {
    int leftPointer = 1;
    int rightPointer = n;
    
    while (leftPointer <= rightPointer) {
        int middlePointer = (rightPointer + leftPointer) / 2;
        
        if (isBadVersion(middlePointer))
            rightPointer = middlePointer - 1;

        else
            leftPointer = middlePointer + 1;
    }

    return rightPointer + 1;
}
